### ISSUE TITLE
GoServer is not able to update materials after it was upgraded to version 24.3.0-19261

### ISSUE THREAD

**Issue by busman76** [gocd/gocd#13147]
##### Issue Type

- Bug Report

##### Summary

GoServer is not able to update materials after it was upgraded to version 24.3.0-19261. We have TFS and GIT materials (both are using https repositories).

The upgrade overwrited file _cacerts_ in \jre\lib\security\ and when GoServer started again (after GoServer was upgraded) we have this error in go-server-wrapper.log: 

`Exception in thread "140@MessageListener for MaterialUpdateListener" java.lang.Error: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`

Similar error is in go-server.log when we make test-connection at configuring a TFS material:

`handleException /go/api/admin/internal/material_test javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target`

and there many lines also in go-server.log like these:

`[Material Update] Skipping update of material TfsMaterial{url=https://xxx/yyy, domain='', projectPath=$/aaa/bbb} which has been in-progress since Wed Oct 16 15:44:48 CEST 2024`

`[Material Update] Skipping update of material GitMaterial{url=https://hhh/jjj, branch='master', shallowClone=false, submoduleFolder='null'} which has been in-progress since Wed Oct 16 15:44:48 CEST 2024`

Of course, we tried to fix file _cacerts_ (put the necessary trusted certificates to it) and after restarted the GoServer we have this error in go-server-wrapper.log: 

`Exception in thread "145@MessageListener for MaterialUpdateListener" java.lang.Error: javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty`

And the error in go-server.log also was change to similar (when we make test-connection at configuring a TFS material):

`handleException /go/api/admin/internal/material_test javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty`

And there are lines in go-server.log which are described above (`[Material Update] Skipping update...`)

We tried to modify the password of file _cacerts_ to _changeit_ or to _serverKeystorepa55w0rd_ - it did not help. Seems to that if we modify the original file _cacerts_ (either putting trusted certificates in it or changing it's password) then we always have the issue "_the trustAnchors parameter must be non-empty_". With the original cacerts (coming with the installation package) we have the issue "_unable to find valid certification path to requested target_".

Actually we are stucked: we are not able to make any build because of this issue.

###### Basic environment details

* Go Version: 24.3.0-19261
* JAVA Version: 21.0.4
* OS: Windows Server 2019(10.0)

##### Log snippets
```
INFO   | jvm 1    | 2024/10/16 15:44:53 | Exception in thread "140@MessageListener for MaterialUpdateListener" java.lang.Error: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1071)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.print(HttpConnection.java:1186)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.MultiThreadedHttpConnectionManager$HttpConnectionAdapter.print(MultiThreadedHttpConnectionManager.java:1793)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.writeRequestLine(HttpMethodBase.java:2383)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.writeRequest(HttpMethodBase.java:2208)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.execute(HttpMethodBase.java:1202)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.java:452)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:189)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpClient.executeMethod(HttpClient.java:428)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpClient.executeMethod(HttpClient.java:343)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.ws.runtime.client.SOAPService.executeSOAPRequestInternal(SOAPService.java:533)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.ws.runtime.client.SOAPService.executeSOAPRequest(SOAPService.java:437)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at ms.ws._LocationWebServiceSoap12Service.connect(_LocationWebServiceSoap12Service.java:385)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.framework.location.internal.LocationWebServiceProxy.connect(LocationWebServiceProxy.java:58)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.connect(LocationService.java:719)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.ensureConnected(LocationService.java:835)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.getInstanceID(LocationService.java:890)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.TFSConnection.getInstanceID(TFSConnection.java:545)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.versioncontrol.VersionControlClient.<init>(VersionControlClient.java:456)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.clients.versioncontrol.VersionControlClient.<init>(VersionControlClient.java:516)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.config.client.DefaultClientFactory$7.newClient(DefaultClientFactory.java:222)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.config.client.DefaultClientFactory.newClient(DefaultClientFactory.java:86)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.TFSConnection.getClient(TFSConnection.java:1490)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.TFSTeamProjectCollection.getVersionControlClient(TFSTeamProjectCollection.java:376)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.tfssdk14.TfsSDKCommand.init(TfsSDKCommand.java:176)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.tfssdk.TfsSDKCommandTCLAdapter.modificationsSince(TfsSDKCommandTCLAdapter.java:75)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.config.materials.tfs.TfsMaterial.modificationsSince(TfsMaterial.java:170)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.service.materials.TfsPoller.modificationsSince(TfsPoller.java:35)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.service.materials.TfsPoller.modificationsSince(TfsPoller.java:26)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.service.MaterialService.modificationsSince(MaterialService.java:135)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:56)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.insertLatestOrNewModifications(MaterialDatabaseUpdater.java:157)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterialWithNewRevisions(MaterialDatabaseUpdater.java:149)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$2.doInTransaction(MaterialDatabaseUpdater.java:108)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:23)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.transaction.TransactionTemplate.lambda$executeWithExceptionHandling$1(TransactionTemplate.java:43)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:40)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:105)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:64)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:32)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:83)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:63)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/java.lang.Thread.run(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | Caused by: javax.net.ssl.SSLHandshakeException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLHandshake.consume(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.TransportContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLTransport.decode(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLSocketImpl.decode(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLSocketImpl.ensureNegotiated(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1109)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1062)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	... 43 more
INFO   | jvm 1    | 2024/10/16 15:44:53 | Caused by: sun.security.validator.ValidatorException: PKIX path building failed: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.validator.PKIXValidator.doBuild(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.validator.PKIXValidator.engineValidate(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.validator.Validator.validate(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at com.microsoft.tfs.core.config.httpclient.internal.DefaultX509TrustManager.checkServerTrusted(DefaultX509TrustManager.java:164)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.ssl.AbstractTrustManagerWrapper.checkServerTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	... 58 more
INFO   | jvm 1    | 2024/10/16 15:44:53 | Caused by: sun.security.provider.certpath.SunCertPathBuilderException: unable to find valid certification path to requested target
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.provider.certpath.SunCertPathBuilder.build(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/sun.security.provider.certpath.SunCertPathBuilder.engineBuild(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	at java.base/java.security.cert.CertPathBuilder.build(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:44:53 | 	... 65 more
```

```
INFO   | jvm 1    | 2024/10/16 15:27:42 | Exception in thread "145@MessageListener for MaterialUpdateListener" java.lang.Error: javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1071)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.print(HttpConnection.java:1186)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.MultiThreadedHttpConnectionManager$HttpConnectionAdapter.print(MultiThreadedHttpConnectionManager.java:1793)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.writeRequestLine(HttpMethodBase.java:2383)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.writeRequest(HttpMethodBase.java:2208)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpMethodBase.execute(HttpMethodBase.java:1202)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpMethodDirector.executeWithRetry(HttpMethodDirector.java:452)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpMethodDirector.executeMethod(HttpMethodDirector.java:189)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpClient.executeMethod(HttpClient.java:428)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpClient.executeMethod(HttpClient.java:343)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.ws.runtime.client.SOAPService.executeSOAPRequestInternal(SOAPService.java:533)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.ws.runtime.client.SOAPService.executeSOAPRequest(SOAPService.java:437)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at ms.ws._LocationWebServiceSoap12Service.connect(_LocationWebServiceSoap12Service.java:385)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.framework.location.internal.LocationWebServiceProxy.connect(LocationWebServiceProxy.java:58)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.connect(LocationService.java:719)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.ensureConnected(LocationService.java:835)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.framework.location.LocationService.getInstanceID(LocationService.java:890)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.TFSConnection.getInstanceID(TFSConnection.java:545)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.versioncontrol.VersionControlClient.<init>(VersionControlClient.java:456)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.clients.versioncontrol.VersionControlClient.<init>(VersionControlClient.java:516)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.config.client.DefaultClientFactory$7.newClient(DefaultClientFactory.java:222)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.config.client.DefaultClientFactory.newClient(DefaultClientFactory.java:86)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.TFSConnection.getClient(TFSConnection.java:1490)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.TFSTeamProjectCollection.getVersionControlClient(TFSTeamProjectCollection.java:376)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.tfssdk14.TfsSDKCommand.init(TfsSDKCommand.java:176)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.tfssdk.TfsSDKCommandTCLAdapter.modificationsSince(TfsSDKCommandTCLAdapter.java:75)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.config.materials.tfs.TfsMaterial.modificationsSince(TfsMaterial.java:170)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.service.materials.TfsPoller.modificationsSince(TfsPoller.java:35)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.service.materials.TfsPoller.modificationsSince(TfsPoller.java:26)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.service.MaterialService.modificationsSince(MaterialService.java:135)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.ScmMaterialUpdater.insertLatestOrNewModifications(ScmMaterialUpdater.java:56)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.insertLatestOrNewModifications(MaterialDatabaseUpdater.java:157)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterialWithNewRevisions(MaterialDatabaseUpdater.java:149)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater$2.doInTransaction(MaterialDatabaseUpdater.java:108)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.transaction.TransactionCallback.doWithExceptionHandling(TransactionCallback.java:23)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.transaction.TransactionTemplate.lambda$executeWithExceptionHandling$1(TransactionTemplate.java:43)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:133)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.transaction.TransactionTemplate.executeWithExceptionHandling(TransactionTemplate.java:40)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialDatabaseUpdater.updateMaterial(MaterialDatabaseUpdater.java:105)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:64)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.materials.MaterialUpdateListener.onMessage(MaterialUpdateListener.java:32)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.runImpl(JMSMessageListenerAdapter.java:83)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.thoughtworks.go.server.messaging.activemq.JMSMessageListenerAdapter.run(JMSMessageListenerAdapter.java:63)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/java.lang.Thread.run(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | Caused by: javax.net.ssl.SSLException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.Alert.createSSLException(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.TransportContext.fatal(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl.handleException(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl.startHandshake(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl.ensureNegotiated(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl$AppOutputStream.write(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1109)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.httpclient.HttpConnection.write(HttpConnection.java:1062)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	... 43 more
INFO   | jvm 1    | 2024/10/16 15:27:42 | Caused by: java.lang.RuntimeException: Unexpected error: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.validator.PKIXValidator.<init>(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.validator.Validator.getInstance(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.getValidator(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrustedInit(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.X509TrustManagerImpl.checkServerTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at com.microsoft.tfs.core.config.httpclient.internal.DefaultX509TrustManager.checkServerTrusted(DefaultX509TrustManager.java:164)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.AbstractTrustManagerWrapper.checkServerTrusted(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.checkServerCerts(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.onCertificate(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.CertificateMessage$T12CertificateConsumer.consume(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLHandshake.consume(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.HandshakeContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.TransportContext.dispatch(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLTransport.decode(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl.decode(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/sun.security.ssl.SSLSocketImpl.readHandshakeRecord(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	... 48 more
INFO   | jvm 1    | 2024/10/16 15:27:42 | Caused by: java.security.InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/java.security.cert.PKIXParameters.setTrustAnchors(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/java.security.cert.PKIXParameters.<init>(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	at java.base/java.security.cert.PKIXBuilderParameters.<init>(Unknown Source)
INFO   | jvm 1    | 2024/10/16 15:27:42 | 	... 66 more
```

**Comment by sriramnrn** [comment]
Are you able to use the openssl s_client to use the certificate file and
connect to the TFS endpoint? I do this with custom CA certificates on my
current project.

A google search of the error
"java.security.InvalidAlgorithmParameterException: the trustAnchors
parameter must be non-empty" led me to
https://support.neo4j.com/s/article/12939005342995-Resolving-SSL-exception-TrustAnchors-parameter-must-be-non-empty

**Comment by chadlwilson** [comment]
I assume from your message that your TFS/TFVC server uses an internally issued certificate, and you need to tell GoCD how to trust the CA that issues its cert (or trust the server cert directly if self-signed cert).

`InvalidAlgorithmParameterException: the trustAnchors parameter must be non-empty`

This likely means the JVM truststore is empty, corrupt, unable to be opened, or unable to be found at all.

1. What was the previous GoCD version and JVM version you were using that was working fine with your TFS?
2. Double-check the permissions on the `cacerts` file if you have manually edited it to make sure the user you run GoCD with can still read the file.
3. Do you/did you previously override any JVM system properties within `config/wrapper-properties.conf` or `config/wrapper.conf`?
    - A common mistake people make is to modify or set custom properties inside `wrapper.conf` in a way that will be overridden by GoCD on an upgrade, rather than retained as for `wrapper-properties.conf`. Might want to check a config backup.
    - You might want to check if you had previously overridden `javax.net.ssl.trustStore`/`trustStorePassword`/`trustStoreType` or similar system properties, but have lost these settings somehow. Generally you should not change the default password of the built-in `cacerts`.
4. Did you try just using the previous `cacerts` file that was working by copying it over the packaged `cacerts` wholesale?
5. does `keytool -list -keystore <path_to_cacerts>` with password `changeit` list the cacerts and your custom cert correctly and consistently as trusted certs?
6. are you sure you are importing the cert correctly as a trusted cert, using `keytool` correctly (i.e including the important `-trustcacerts` flag for keytool?
7. you could try getting much more detailed debugging information by setting `-Djavax.net.debug=ssl` in your `wrapper-properties.conf`, however can't recall if it'll log the truststore/cacerts load process
8. You probably also have agents that need to connect to the same TFS server and would need the same certificate to be trusted by the agent JVM. If you have agents running somewhere that you haven't yet upgraded, perhaps take a look at their `cacerts` or properties to find how it was working before. (assuming you did not

To avoid such overwrites in future, good practice would probably be to set `wrapper.java.additional.100=-Djavax.net.ssl.trustStore=\path\to\custom\cacerts` in `wrapper-properties.conf` so you are not modifying files that the installer owns such as the JRE `cacerts` (but note that you'd be taking ownership for ensuring cacerts get updated periodically)

**Comment by busman76** [comment]
We fixed the issue - the problem was caused by the `keytool`. In the questioned windows server (where the GoServer is running) we started the `keytool` from command line and its pointed to an older version of `keytool` (from `c:\Program Files (x86)\Java\jre1.8.0_291\bin\`). So the newly created `cacerts `was incompatible with GoCD. After we changed `keytool` path to `c:\Program Files (x86)\Go Server\jre\bin\` and put the trusted certificates to `cacerts`, the `trustAnchors` issue was gone. And, seems to that we do not have to use password for `cacerts` (how is it possible?).

Anyhow, thanks for all the hints you wrote - we are going to review our settings.
